internal/app/networkd/pkg/reg/reg.go

```diff
@@ -155,16 +156,54 @@ func toCIDR(family uint8, prefix net.IP, prefixLen int) string {
 }
 
 // Check implements the Health api and provides visibilty into the state of networkd.
-// Ready signifies the daemon (api) is healthy and ready to serve requests.
+// We determine networkd health based on neighbor (arp table) entries.
+// Under normal operating circumstances, there should be a steady state of reachable neighbor entries.
+// If we get in a situation where all of our neighbor entries are stale, then we're in trouble.
 func (r *Registrator) Check(ctx context.Context, in *empty.Empty) (reply *healthapi.HealthCheckResponse, err error) {
+	// Set initial state to unknown
 	reply = &healthapi.HealthCheckResponse{
 		Messages: []*healthapi.HealthCheck{
 			{
-				Status: healthapi.HealthCheck_SERVING,
+				Status: healthapi.HealthCheck_UNKNOWN,
 			},
 		},
 	}
 
+	var neighbors []rtnetlink.NeighMessage
+
+	neighbors, err = r.Conn.Neigh.List()
+	if err != nil {
+		return reply, err
+	}
+
+	// After getting a list of neighbors we can upgrade to not serving
+	reply.Messages[0].Status = healthapi.HealthCheck_NOT_SERVING
+
+	// Find at least one neighbor in a reachable state
+	for _, neighbor := range neighbors {
+		// Verify neighbor is associated with a link
+		// managed by networkd ( skip cni links )
+		link, err := r.Conn.Link.Get(neighbor.Index)
+		if err != nil {
+			continue
+		}
+
+		if _, ok := r.Networkd.Interfaces[link.Attributes.Name]; !ok {
```

Comment:
I guess this map might be mutated while API is running and at the same time networkd is discovering new interfaces?

so potential data race here, with map it's even worse potentially

Reply:
I'm pretty sure this should be fine with the given workflow. networkd.Interfaces only gets populated during the `Configure()` phase and since we're only doing key lookups ( we don't care about the value ) I'm not sure I see there being any harm this way. 
There's probably another ticket worth to potentially refactor it from a map to something more useful, so maybe something for a follow up pr.

Reply:
yep, we start actually API server when initial configuration is done: https://github.com/talos-systems/talos/blob/master/internal/app/networkd/main.go#L46-L60

so this shoudl be safe.